### DISCUSSION TITLE
DNS response is different when I use my os dns and when I use CoreDNS

### DISCUSSION THREAD

**Discussion by juchom** [coredns/coredns#5971] [Q&A]
Hello,

I have a domain I use for local development with two records and one record for let's encrypt validation :
```
@                  A    127.0.0.1
*                  A    127.0.0.1
_acme-challenge   TXT   xxxxxxxxxxxxxxxxxxxxx
```

I have an error with cert-manager complaining about a recursive CNAME when it runs in Kubernetes but works fine when I run it outside Kubernetes.

If i run this command outside the cluster I have this result (the expected one)

```
dig cname _acme-challenge.domain.dev.

; <<>> DiG 9.18.1-1ubuntu1.2-Ubuntu <<>> cname _acme-challenge.domain.dev.
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 30141
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1472
;; QUESTION SECTION:
;_acme-challenge.domain.dev.     IN      CNAME

;; AUTHORITY SECTION:
domain.dev.              78      IN      SOA     desi.ns.cloudflare.com. dns.cloudflare.com. 2304297281 10000 2400 604800 3600

;; Query time: 0 msec
;; SERVER: 172.28.208.1#53(172.28.208.1) (UDP)
;; WHEN: Tue Mar 14 18:02:22 CET 2023
;; MSG SIZE  rcvd: 116
```

If I run the same command inside the Kubernetes cluster I have this weird response with a CNAME pointing to the dig query :

```
dig cname _acme-challenge.domain.dev.

; <<>> DiG 9.9.5-9+deb8u19-Debian <<>> cname _acme-challenge.domain.dev.
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 34937
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;_acme-challenge.domain.dev.     IN      CNAME

;; ANSWER SECTION:
_acme-challenge.domain.dev. 5    IN      CNAME   _acme-challenge.domain.dev.

;; Query time: 3 msec
;; SERVER: 10.96.0.10#53(10.96.0.10)
;; WHEN: Tue Mar 14 17:01:
```

Do you have any idea why the resolution differs from the OS and from CoreDNS ?

I'm using docker desktop 4.17.0 on windows 11, with Kubernetes 1.25.4 and CoreDNS-1.9.3

**Comment by chrisohaver** [comment]
This is not a normal behavior in a standard configuration.
Can you share the contents of your Corefile (in the coredns configmap)?

**Comment by juchom** [comment]
Hello @chrisohaver,

Here is the config map : `kubectl -n kube-system edit configmap coredns`

```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
data:
  Corefile: |
    .:53 {
        errors
        health {
           lameduck 5s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
           pods insecure
           fallthrough in-addr.arpa ip6.arpa
           ttl 30
        }
        prometheus :9153
        forward . /etc/resolv.conf {
           max_concurrent 1000
        }
        cache 30
        loop
        reload
        loadbalance
    }
kind: ConfigMap
metadata:
  creationTimestamp: "2023-03-16T08:59:12Z"
  name: coredns
  namespace: kube-system
  resourceVersion: "233"
  uid: b3f8fa68-347e-4bd4-baa5-cb0fc86eee61

```

**Reply by chrisohaver** [reply]
Hmm, nothing in there out of the ordinary that would be causing that.

Per the config, the CNAME is coming from from the upstream server in /etc/resolv.conf, the contents of which are controlled by kubernetes (kublet).  This is usually (by default) set to be the same as the /etc/resolv.conf on the node the coredns pod is running on.

From where are you doing the "outside of kubernetes" query?  If not the kubernetes node, are the contents of /etc/resolv.conf the same as on the kubernetes node?

**Reply by juchom** [reply]
It's on the same machine, I'm using the docker desktop kubernetes cluster.

The dig command is made from the WSL2.

**Reply by juchom** [reply]
I also tried with minikube and the results are the same.

**Reply by juchom** [reply]
Here is the resolv.conf from the host machine :

```
cat /etc/resolv.conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver 172.31.80.1
```

Here is the one from Kuberenetes (I followed instructions from here : https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/#check-the-local-dns-configuration-first)

```
kubectl exec -ti dnsutils -- cat /etc/resolv.conf
nameserver 10.96.0.10
search default.svc.cluster.local svc.cluster.local cluster.local
options ndots:5
```

**Reply by chrisohaver** [reply]
Is your host machine the node? or are your nodes docker images (or otherwise virtual)?  

If the latter, then there are separate 4 things:
1) the host machine (windows 11 in your case) - I think this is where you made the "out of kubernetes" query from?
2) the kubernetes node. Can you make a dig query from this node?
3) the coredns pods
4) other pods (which use clusterfirst dnspolicy)

The coredns pod will use the same /etc/resolv.conf as the kubernetes node.

other pods will use a specialized "clusterfirst"  resolv.conf that is geared for k8s service discovery. That's what you see in your second one above.

**Reply by juchom** [reply]
@chrisohaver I'm pretty new to kubernetes and I'm not sure on how to answer this question : `Is your host machine the node? or are your nodes docker images (or otherwise virtual)?`

I'll try to describe you my environment : I have a laptop, running windows 11 & WSL 2 Ubuntu 22.04, docker desktop for windows 4.17.0.
I activated the Kubernetes cluster offered by docker for desktop. There is nothing more, I suppose that the cluster is just one pod hosted on my machine.

`the host machine (windows 11 in your case) - I think this is where you made the "out of kubernetes" query from?` YES

The result of the dig query is this one :

```
dig cname _acme-challenge.domain.dev.

; <<>> DiG 9.18.1-1ubuntu1.2-Ubuntu <<>> cname _acme-challenge.domain.dev.
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 30141
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1472
;; QUESTION SECTION:
;_acme-challenge.domain.dev.     IN      CNAME

;; AUTHORITY SECTION:
domain.dev.              78      IN      SOA     desi.ns.cloudflare.com. dns.cloudflare.com. 2304297281 10000 2400 604800 3600

;; Query time: 0 msec
;; SERVER: 172.28.208.1#53(172.28.208.1) (UDP)
;; WHEN: Tue Mar 14 18:02:22 CET 2023
;; MSG SIZE  rcvd: 116
```

And running the same command from the dnsutils pod :

```
kubectl exec -ti dnsutils -- dig cname _acme-challenge.domain.dev.

; <<>> DiG 9.9.5-9+deb8u19-Debian <<>> cname _acme-challenge.domain.dev.
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 21977
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;_acme-challenge.domain.dev.     IN      CNAME

;; ANSWER SECTION:
_acme-challenge.domain.dev. 5    IN      CNAME   _acme-challenge.domain.dev.

;; Query time: 4 msec
;; SERVER: 10.96.0.10#53(10.96.0.10)
;; WHEN: Thu Mar 16 19:00:41 UTC 2023
;; MSG SIZE  rcvd: 118
```

**Comment by chrisohaver** [comment]
per https://docs.docker.com/desktop/kubernetes/,

"The Kubernetes server runs locally within your Docker instance, is not configurable, and is a single-node cluster. It runs within a Docker container on your local system, and is only for local testing."

That means the k8s node is not the same system as your host, it's a separate system, with it's own /etc/resolv.conf, the content of which may be different than the host's /etc/resolv.conf.  Therefore DNS queries could be sent to different DNS servers (and getting different answers).

To verify if thats the case, you need to figure out what the /etc/resolv.conf is on the k8s node.

To see what the /etc/resolv.conf is on the k8s node, you can either:
* get a shell into the docker image node and cat it, if allows you to do that.
* if you cant do that, then you can create a new pod in the cluster with dnsPolicy = "default" (same a s coredns pod).  This will create a pod that has the same /etc/resolv.conf as the node. Then open a shell in that pod and examine /etc/resolv.conf.

**Reply by juchom** [reply]
Hello,

Sorry for the late response, there is a bunch of images to make kubernetes running, but here is the content of `/etc/resolv.conf` from this image `k8s_coredns_coredns-565d847f94-m7984_kube-system_c57bf7cd-d6c4-4ff9-b927-a4104e6ceb08_2`

Inside the pod (my computer is on `192.168.0.0/24`)

```
nameserver 192.168.65.7
```

The one from this image is `k8s_POD_vpnkit-controller_kube-system_51f6b418-60f9-440a-8b0d-be64b327201d_8`:

```
nameserver 10.96.0.10
search kube-system.svc.cluster.local svc.cluster.local cluster.local
options ndots:5
```

Inside WSL / Windows (my isp)

```
nameserver 172.31.80.1
```

Thanks again for your support this very nice 👍

**Reply by juchom** [reply]
Ok, this issue is linked to docker desktop for windows. I tried the same command with docker desktop for linux and it behaves normally.

**Reply by chrisohaver** [reply]
It is an odd issue that I have never seen before.  My only experience with Docker Desktop is on Mac, and I'll admit I have not used its implementation of Kubernetes.  I just use kind.
